HB/structure.elpi

```diff
@@ -276,6 +284,47 @@ export-operations Structure ProjSort ProjClass MLwP EX1 EX2 MLToExport :- std.do
   std.map LMwPToExport w-params_1 MLToExport,
 ].
 
+pred mk-sort-coercion-aux i:term, i:term, i:term, i:list term, o:prop.
+mk-sort-coercion-aux (prod _N _T Body) Structure P Args Clause :-
+  Clause = (pi x\ C x),
+  pi x\ mk-sort-coercion-aux (Body x) Structure P [x | Args] (C x).
+
+mk-sort-coercion-aux _ Structure P Args Clause :-
+  std.rev Args ArgsRev,
+  Clause =
+    (pi ctx v t e r argsAll w\ (coercion ctx v t e r :-
+      coq.unify-eq t (app [Structure | ArgsRev]) ok,
+      std.append ArgsRev [v] argsAll,
+      mk-app P argsAll w,
+      coq.elaborate-skeleton w e r ok,
+      coq.ltac.collect-goals r [] [])).
+
+pred mk-sort-coercion i:term, i:term, o:prop.
+mk-sort-coercion Structure P Clause :-
+  coq.typecheck Structure T ok,
+  mk-sort-coercion-aux T Structure P [] Clause.
+
+pred mk-generic-coercion-aux i:term, i:term, i:term, i:list term, o:prop.
+mk-generic-coercion-aux (prod _N _T Body) Structure G Args Clause :-
+  Clause = (pi x\ C x),
+  pi x\ mk-generic-coercion-aux (Body x) Structure G [x | Args] (C x).
+
+mk-generic-coercion-aux _ Structure G Args Clause :-
+  std.rev Args ArgsRev,
+  Clause =
+    (pi ctx v t e r c argsAll w\ (coercion ctx v t e r :-
+      coq.unify-eq e (app [Structure | ArgsRev]) ok,
+      std.append ArgsRev [v, c] argsAll,
+      mk-app G argsAll w,
+      coq.elaborate-skeleton w e r ok,
+      coq.ltac.collect-goals r [] [])).
+
+pred mk-generic-coercion i:gref, o:prop.
+mk-generic-coercion Structure Clause :-
+  coq.typecheck (global Structure) T ok,
+  get-constructor Structure G,
+  mk-generic-coercion-aux T (global Structure) (global G) [] Clause.
```

Comment:
I'm not so sure I get the name.  Aren't you trying to build a sort of reverse coercion from a value (v : t) to a structure s by building `build_s v ?c` and hoping coq infers the class `?c`?

Reply:
This is exactly what I do. Do you mean `mk-reverse-coercion` would be a better name?